targets/TARGET_ublox/TARGET_R5030/spi_api.c

```diff
@@ -0,0 +1,234 @@
+                                spi_free:
+                          Release a SPI object
+*****************************************************************************/
+void spi_free(spi_t *obj)
+{
+	obj->reg_base->cr |= SPI_CR_CR_ENABLE_DISABLE_VALUE << SPI_CR_CR_ENABLE_OFFSET; // Disable SPI 
+}
+	
+/*****************************************************************************
+                                spi_format:
+                          Configure the SPI format
+*****************************************************************************/
+void spi_format(spi_t *obj, int bits, int mode, int slave)
+{
+    if (slave == 0) {
+        obj->reg_base->cr |= SPI_CR_CR_MASTER_MASTER_MODE_VALUE << SPI_CR_CR_MASTER_OFFSET; // Master Mode Selected
+    } else {
+        obj->reg_base->cr |= SPI_CR_CR_MASTER_SLAVE_MODE_VALUE << SPI_CR_CR_MASTER_OFFSET; // Slave Mode Selected
+    }
+	
+	// only 8 and 9 bit word length supported, 9-bit mode (1 status bit + 8 data bits) to implement flow control
+    MBED_ASSERT(bits == 8 || bits == 9);
+	
+    if (bits == 8) {
+        obj->reg_base->cr |= SPI_CR_CR_FLOWCTRL_NORMAL_OPERATION_VALUE << SPI_CR_CR_FLOWCTRL_OFFSET;
+    } else {
+        obj->reg_base->cr |= SPI_CR_CR_FLOWCTRL_FLOW_CONTROL_9_BIT_MODE_ENABLED_VALUE << SPI_CR_CR_FLOWCTRL_OFFSET;
+    }
+		
+    switch (mode) {
+        case 0:
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPOL_SCLK_LOW_WHEN_IDLE_VALUE << SPI_CR_CR_CLOCKPOL_OFFSET;
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPHA_FIRST_DATA_VALID_ON_FIRST_SCLK_EDGE_VALUE << SPI_CR_CR_CLOCKPHA_OFFSET;
+            break;
+        case 1:
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPOL_SCLK_LOW_WHEN_IDLE_VALUE << SPI_CR_CR_CLOCKPOL_OFFSET;
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPHA_FIRST_DATA_VALID_ON_SECOND_SCLK_EDGE_VALUE << SPI_CR_CR_CLOCKPHA_OFFSET;
+            break;
+        case 2:
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPOL_SCLK_HIGH_WHEN_IDLE_VALUE << SPI_CR_CR_CLOCKPOL_OFFSET;
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPHA_FIRST_DATA_VALID_ON_FIRST_SCLK_EDGE_VALUE << SPI_CR_CR_CLOCKPHA_OFFSET;
+            break;
+        case 3:
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPOL_SCLK_HIGH_WHEN_IDLE_VALUE << SPI_CR_CR_CLOCKPOL_OFFSET;
+            obj->reg_base->cr |= SPI_CR_CR_CLOCKPHA_FIRST_DATA_VALID_ON_SECOND_SCLK_EDGE_VALUE << SPI_CR_CR_CLOCKPHA_OFFSET;
+            break;
+        default:
+            break;
+    }
+}
+
+/*****************************************************************************
+                                spi_frequency:
+                            Set the SPI baud rate
+*****************************************************************************/
+void spi_frequency(spi_t *obj, int hz)
+{
+    uint8_t clockDiv = 0, p_waitDiv = 0;
+	
+    if (bit_status(obj->reg_base->cr, SPI_CR_CR_MASTER_OFFSET)) {
+        MBED_ASSERT(hz <= MAX_FREQ); // The maximum bus clock speed in Master mode shall be 26 MHz
```

Comment:
I don't think there is any limitation to minimum frequency. The **hz** is the required SPI frequency(SCK) not the bus frequency. In master mode, SCK can be upto 1/2 times the bus frequency and in slave mode SCK can be upto 1/6 times the bus frequency. So for FPGA, in master mode, maximum frequency is 13MHz and in slave mode, the maximum frequency is 4.33MHz. 

One more thing, Guanhua mentioned that the FPGA cannot keep up with a high speed SPI so the FiFo starts overflowing. I suggest you hard code frequency to be 50KHz (which has been tested reliably both by me and Guanhua) with a comment that it is done only for testing on FPGA and will later be removed.

Reply:
It is mention in Klein Matterhorn Target Specification document that the maximum bus clock speed in Master mode shall be 26 MHz , the max. bus-clock speed in Slave mode shall not be less than 13 MHz but I think these are hardware requirement only, will change it as mention in SPI document. 

Regarding the tested frequency of 50KHz, I think we should set it through application, not hard coded for our testing as well.